### PR TITLE
#58: Added type hints

### LINKED ISSUE #58: Add Python type hints
- we need to add type hints to functions
- eg. https://github.com/exasol/bucketfs-utils-python/blob/56a2db5ecc4098d53f2dc562ff91fda33b32e1b7/exasol_bucketfs_utils_python/bucketfs_location.py#L57-L59

### PR DESCRIPTION
**Pull request by umitbuyuksahin**
Fixes #58

### CODE REVIEW

File: exasol_bucketfs_utils_python/abstract_bucketfs_location.py [hunk 1]
```diff
@@ -1,52 +1,67 @@
...
     @abstractmethod
     def upload_fileobj_to_bucketfs(self,
-                                   fileobj: typing.IO,
-                                   bucket_file_path: str):
+                                   fileobj: IO,
+                                   bucket_file_path: str) -> \
+            Tuple[ParseResult, PurePosixPath]:
         pass
 
     # TODO add missing upload/download functions
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/abstract_bucketfs_location.py:45]: We don't want TODOS in merged code as far as i know. Make a ticket for adding these instead.

File: exasol_bucketfs_utils_python/bucketfs_connection_config.py [hunk 2]
```diff
@@ -8,14 +8,16 @@ class BucketFSConnectionConfig:
     """
 
     @typechecked(always=True)
-    def __init__(self, host: str, port: int, user: str, pwd: str, is_https=False):
+    def __init__(self, host: str, port: int,
+                 user: str, pwd: str, is_https=False):
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/bucketfs_connection_config.py:12]: is_https has no typehint
```suggestion
                 user: str, pwd: str, is_https: str =False):
```

File: exasol_bucketfs_utils_python/bucketfs_factory.py [hunk 3]
```diff
@@ -1,25 +1,29 @@
...
+    LocalFSMockBucketFSLocation
 
 
 class BucketFSFactory:
     """
     Creates a BucketFSLocation given an url.
     """
-    def create_bucketfs_location(self, url: str, user: str, pwd: str, base_path: Optional[PurePosixPath] = None):
+    def create_bucketfs_location(self, url: str, user: str, pwd: str,
+                                 base_path: Optional[PurePosixPath] = None):
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/bucketfs_factory.py:18]: missing return type hint

File: exasol_bucketfs_utils_python/bucketfs_utils.py [hunk 4]
```diff
@@ -16,26 +14,30 @@ def _encode_url_part(part: str) -> str:
...
 
 
 @typechecked(always=True)
-def generate_bucketfs_udf_path(bucketfs_config: BucketFSConfig) -> PurePosixPath:
+def generate_bucketfs_udf_path(bucketfs_config: BucketFSConfig) \
+        -> PurePosixPath:
     """
-    This function generates the path where UDFs can access the content of a BucketFS in their file system
+    This function generates the path where UDFs can access the content of a
+    BucketFS in their file system
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/bucketfs_utils.py:40]:
```suggestion
    BucketFS in their file system.
```

File: exasol_bucketfs_utils_python/bucketfs_utils.py [hunk 5]
```diff
@@ -45,11 +47,12 @@ def generate_bucketfs_udf_path(bucketfs_config: BucketFSConfig) -> PurePosixPath
...
-def generate_bucket_udf_path(bucket_config: BucketConfig,
-                             path_in_bucket: Union[None, str, PurePosixPath]) -> PurePosixPath:
+def generate_bucket_udf_path(
+        bucket_config: BucketConfig,
+        path_in_bucket: Union[None, str, PurePosixPath]) -> PurePosixPath:
     """
-    This function generates the path where UDFs can access the content of a bucket or
-    the given path in a bucket in their file system
+    This function generates the path where UDFs can access the content of a
+    bucket or the given path in a bucket in their file system
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/bucketfs_utils.py:55]:
```suggestion
    bucket or the given path in a bucket in their file system.
```

File: exasol_bucketfs_utils_python/download.py [hunk 6]
```diff
@@ -1,18 +1,19 @@
...
-
 from exasol_bucketfs_utils_python import bucketfs_utils
 from exasol_bucketfs_utils_python.bucket_config import BucketConfig
 from exasol_bucketfs_utils_python.bucketfs_utils import generate_bucket_http_url
 
 
-def download_from_bucketfs_to_file(bucket_config: BucketConfig, bucket_file_path: str, local_file_path: Path):
+def download_from_bucketfs_to_file(bucket_config: BucketConfig,
+                                   bucket_file_path: str,
+                                   local_file_path: Path) -> None:
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/download.py:13]: For some functions which do not return anything you have no return types, and for others you have return type "None". Maybe make it more consistent in that regard?

File: exasol_bucketfs_utils_python/download.py [hunk 7]
```diff
@@ -1,18 +1,19 @@
...
     """
-    Download a file from the specified path in the bucket in the BucketFs and save as a local file
+    Download a file from the specified path in the bucket in the BucketFs
+    and save as a local file
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/download.py:16]:
```suggestion
    and save as a local file.
```

File: exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py [hunk 8]
```diff
@@ -5,14 +5,15 @@
 
 
 class GithubReleaseFileBucketFSUploader:
-    def __init__(self, file_to_download_name, github_user, repository_name, release_name, path_inside_bucket):
+    def __init__(self, file_to_download_name, github_user,
+                 repository_name, release_name, path_inside_bucket):
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py:9]: Missing type hints

File: exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py [hunk 9]
```diff
@@ -5,14 +5,15 @@
...
         self.file_to_download_name = file_to_download_name
         self.github_user = github_user
         self.repository_name = repository_name
         self.release_name = release_name
         self.path_inside_bucket = path_inside_bucket
 
-    def upload(self, address, username, password):
+    def upload(self, address, username, password) -> None:
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py:16]: Missing type hints

File: exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py [hunk 10]
```diff
@@ -24,9 +25,12 @@ def upload(self, address, username, password):
...
         r_download = requests.get(download_url, stream=True)
         upload_url = self.__build_upload_url(address)
-        requests.put(upload_url, data=r_download.iter_content(10 * 1024), auth=HTTPBasicAuth(username, password))
+        requests.put(
+            upload_url,
+            data=r_download.iter_content(10 * 1024),
+            auth=HTTPBasicAuth(username, password))
 
-    def __build_upload_url(self, address):
+    def __build_upload_url(self, address) -> str:
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py:33]: Missing type hint for address

File: exasol_bucketfs_utils_python/release_link_extractor.py [hunk 11]
```diff
@@ -9,23 +9,27 @@ def __init__(self, repository_api_link):
         """
         self.repository_api_link = repository_api_link
 
-    def get_link_by_release_name(self, file_to_download_name):
+    def get_link_by_release_name(self, file_to_download_name) -> str:
```
Comment (MarleneKress79789) [inline exasol_bucketfs_utils_python/release_link_extractor.py:12]: Missing type hint for file_to_download_name

File: exasol_bucketfs_utils_python/abstract_bucketfs_location.py (hunk 1 above)
Comment (umitbuyuksahin) [reply exasol_bucketfs_utils_python/abstract_bucketfs_location.py:45]: I don't know the details of that comment, better to leave it like that

File: exasol_bucketfs_utils_python/bucketfs_connection_config.py (hunk 2 above)
Comment (umitbuyuksahin) [reply exasol_bucketfs_utils_python/bucketfs_connection_config.py:12]: thx, it is boolean.

File: exasol_bucketfs_utils_python/download.py (hunk 6 above)
Comment (umitbuyuksahin) [reply exasol_bucketfs_utils_python/download.py:13]: I skipped some , all must return None

File: exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py (hunk 10 above)
Comment (MarleneKress79789) [reply exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py:34]: Same here, no type hint for address

Comment (umitbuyuksahin) [reply exasol_bucketfs_utils_python/github_release_file_bucketfs_uploader.py]: thx, fixed now

**Review by MarleneKress79789 — APPROVED** [review]

**Pull request merged**